### PR TITLE
126 fix show the posts in order of publishing date with most recent on top

### LINKED ISSUE #126: Fix: Show the posts in order of publishing date with most recent on top
The actual task was fixed in 30 minutes but I used the opportunity to refactor the code more and make it more maintainable

### PR DESCRIPTION
**Pull request by yasmineben510**
In this PR, I ensured the most recent posts are always at the top of the list, and in the meantime, I also modulated the function to be able to take as a parameter and specify how far ago we should take the posts from.

I also added a test to ensure this bug doesn't resurface

This PR closes #126

### CODE REVIEW

File: app/src/main/java/com/github/orkest/ui/feed/PostViewModel.kt [hunk 1]
```diff
@@ -94,8 +95,9 @@ open class PostViewModel {
     /**
      * Returns a list of the most recent posts since the last connection of the user from the database
      */
-    open fun getRecentPosts(lastConnected: LocalDateTime): CompletableFuture<List<Post>>{
-        return dbAPI.getRecentPostsFromDataBase(lastConnected)
+    open fun getRecentPosts(lastConnected: LocalDateTime,
+                            maxDaysNb: Long = DEFAULT_MAX_RECENT_DAYS ): CompletableFuture<List<Post>>{
+        return dbAPI.getRecentPostsFromDataBase(lastConnected, maxDaysNb)
     }
```
Comment (saraanej) [inline app/src/main/java/com/github/orkest/ui/feed/PostViewModel.kt:101]: I actually don't understand the purpose of giving maxDaysNb as a parameter to the function. It is not something that will be changed depending on the post so it shouldn't stay as a parameter. Only calling the constant MAX_RECENT_DAYS is enough and if we want to change this value, we would know exactly where to do so.

File: app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt [hunk 2]
```diff
@@ -311,6 +299,47 @@ class FireStoreDatabaseAPI {
...
+                listMonths.add(recentPostsTime.monthValue+1)
+
+        //Get all posts that were published after the recentPostsTime - One query per month
+        listMonths.forEach { month ->
+            run {
+                futures.add(recentPostsQuery(month,
+                    recentPostsTime.year,
+                    if (month == recentPostsTime.monthValue) recentPostsTime.dayOfMonth else 1))
+            }
+        }
```
Comment (saraanej) [inline app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt:334]: Why did you need to wrap your code in a run block  ? Did you have an error when not doing so ?

File: app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt [hunk 3]
```diff
@@ -311,6 +299,47 @@ class FireStoreDatabaseAPI {
         return future
     }
 
+
+    //TODO: Add the condition on posts belonging to the user's friends
+    /**
```
Comment (saraanej) [inline app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt:304]: Very good remark. I'm not sure it would be implemented in the function of fetching the posts form the database though but rather when calling the function, we would need to call it only for the user's followings.

File: app/src/main/java/com/github/orkest/data/Constants.kt [hunk 4]
```diff
@@ -1,62 +1,63 @@
...
+import java.time.LocalDateTime
+import java.time.ZoneId
+
+/**
+ * Class to store the current logged-in username since it is needed in test classes, composable functions and viewModels
+ */
+class Constants {
+    companion object{
+
+        const val DEFAULT_MAX_RECENT_DAYS: Long = 30
```
Comment (saraanej) [inline app/src/main/java/com/github/orkest/data/Constants.kt:18]: A more descriptive name could replace this name as it is not very intuitive about what it represents.

File: app/src/androidTest/java/com/github/orkest/ViewModel/feed/PostViewModelTest.kt [hunk 5]
```diff
@@ -70,6 +70,35 @@ class PostViewModelTest {
         assert(posts[0].postDescription == "Recent1")
     }
 
+    @Test
+    fun getRecentPostsReturnsPostsInDescendingOrderOfDate(){
+        Constants.CURRENT_LOGGED_USER = "DummyUser"
+        postViewModel.updateSong(Constants.DUMMY_RUDE_BOY_SONG)
+        postViewModel.updatePostDescription(TextFieldValue("Recent1"))
```
Comment (saraanej) [inline app/src/androidTest/java/com/github/orkest/ViewModel/feed/PostViewModelTest.kt:77]: Very good test. You could maybe add tests about edge cases like when there are no posts to fetch or maxNbDays parameter is higher than the constant MAX_NB_DAYS

**Review by saraanej — APPROVED** [review]
Clean code and very well documented. Thank you very much ! I have some questions about the implementation of the code but it all looks good to me.

File: app/src/main/java/com/github/orkest/ui/feed/PostViewModel.kt (hunk 1 above)
Comment (yasmineben510) [reply app/src/main/java/com/github/orkest/ui/feed/PostViewModel.kt:101]: There could be times where you want to load all posts that are 30 days before and others where you would like to load posts only for the last pas week at the maximum, it is therefore the most maintainable way to do it I think. The MAX_RECENT_DAYS value is actually the maximum allowed value since I am not handling the case for which one would like to load unseen posts more than 1 month old (Made the choice that it would become necessarily complexities with respect to performance and readability)

File: app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt (hunk 2 above)
Comment (yasmineben510) [reply app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt:334]: Yes, without it, I get a warning mentioning that the lambda code was just ignored

File: app/src/main/java/com/github/orkest/ui/feed/PostViewModel.kt (hunk 1 above)
Comment (saraanej) [reply app/src/main/java/com/github/orkest/ui/feed/PostViewModel.kt:101]: What are the cases where we would want to load posts with different time limits ? Is it for caching purposes ?

File: app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt (hunk 3 above)
Comment (yasmineben510) [reply app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt:304]: I am not sure I understood what you mean, the fetching of recents posts is independent of the username right now

Comment (saraanej) [reply app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt:304]: Yes I know, I am making a remark to the TODO you added at the top of the getRecentPostsFromDatabase.

Comment (yasmineben510) [reply app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt:304]: Yes I know, but I didn’t understand what you meant, right now calling the function is not user specific

Comment (saraanej) [reply app/src/main/java/com/github/orkest/domain/FireStoreDatabaseAPI.kt:304]: Because you wrote the TODO at the top of the getRecentPostsFromDatabase, I understood that you wanted to make the function user specific (in a future sprint maybe) to implement the modification. That's why I made that remark :)
